Terraform fails with empty string errors

Since updating from master a week or two ago, Terraform has started giving me mystery errors:

```console
$ awsudo terraform refresh
module.dev_vpc.aws_vpc.vpc: Refreshing state... (ID: vpc-3ac99d5f)
module.internal_vpc.aws_vpc.vpc: Refreshing state... (ID: vpc-6aca9e0f)
module.main_registry_storage.aws_iam_role.rw: Refreshing state... (ID: main-docker-registry-rw)
module.main_registry_storage.aws_iam_role.ro: Refreshing state... (ID: main-docker-registry-ro)
module.main_registry_storage.aws_s3_bucket.registry: Refreshing state... (ID: main-docker-registry-paperg-com)
module.main_registry_storage.aws_iam_instance_profile.rw: Refreshing state... (ID: main-docker-registry-rw)
module.main_registry_storage.aws_iam_instance_profile.ro: Refreshing state... (ID: main-docker-registry-ro)
module.dev_vpc.aws_internet_gateway.gateway: Refreshing state... (ID: igw-bd3ca8d8)
Error refreshing state: 1 error(s) occurred:

* 1 error(s) occurred:

* 1 error(s) occurred:

* 
```

If I run it with `TF_LOG=1` and grep for errors, I see a lot of this:

```
2015/04/17 13:03:09 [ERROR] root.internal_vpc: eval: *terraform.EvalValidateCount, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.internal_vpc: eval: *terraform.EvalOpFilter, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.internal_vpc: eval: *terraform.EvalSequence, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.dev_vpc: eval: *terraform.EvalValidateCount, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.dev_vpc: eval: *terraform.EvalOpFilter, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.dev_vpc: eval: *terraform.EvalSequence, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.main_registry_storage: eval: *terraform.EvalValidateCount, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.main_registry_storage: eval: *terraform.EvalOpFilter, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.main_registry_storage: eval: *terraform.EvalSequence, err: Warnings: []. Errors: []
2015/04/17 13:03:09 [ERROR] root.main_registry_storage: eval: *terraform.EvalValidateCount, err: Warnings: []. Errors: 
```

I'm confused because it says "1 error(s) occurred:" but then there's no error, and the logs say "[ERROR]" and then explain that there's an empty list of errors.

A guess: I recently deleted some resources through the AWS web console, which is necessary because  destroy doesn't work (#922). Running refresh should notice this, and remove the resource from the state file, but it's a common bug for providers to forget about that case (#889). I think I might be hitting such a case somewhere, but without any actual error messages I can only guess.

Sorry I can't narrow it down any more -- I realize this is totally unreproducable. But maybe since this seems like a recent change to master, someone remembers some change that might be relevant.

--- a/builtin/providers/aws/resource_aws_s3_bucket.go
+++ b/builtin/providers/aws/resource_aws_s3_bucket.go
@@ -85,7 +85,13 @@ func resourceAwsS3BucketRead(d *schema.ResourceData, meta interface{}) error {
 		Bucket: aws.String(d.Id()),
 	})
 	if err != nil {
-		return err
+		if awsError, ok := err.(aws.APIError); ok && awsError.StatusCode == 404 {
+			d.SetId("")
+		} else {
+			// some of the AWS SDK's errors can be empty strings, so let's add
+			// some additional context.
+			return fmt.Errorf("error reading S3 bucket \"%s\": %#v", d.Id())
+		}
 	}
 
 	tagSet, err := getTagSetS3(s3conn, d.Id())

